tcmur_cmd_handler.c

```diff
@@ -682,9 +682,19 @@ static int handle_unmap_in_writesame(struct tcmu_device *dev,
 	uint8_t *cdb = cmd->cdb;
 	uint64_t lba = tcmu_cdb_get_lba(cdb);
 	uint64_t nlbas = tcmu_cdb_get_xfer_length(cdb);
+	uint32_t align = tcmu_dev_get_unmap_gran_align(dev);
 	struct unmap_state *state;
 	int ret;
 
+	/* If not aligned then falls back to the writesame without unmap */
+	if (lba % align || nlbas % align) {
+		tcmu_dev_warn(dev,
+			      "Start lba: %"PRIu64" or nlbas: %"PRIu64" not aligns to %"PRIu32"\n",
```

Comment:
`not aligns` -> `not aligned` (also probably not a warning since it's only an issue because `tcmu` is not treating the WS as a WS). Looking at the kernel's `drivers/scsi/sd.c`, it translates a zero extent call to a WS+UNMAP and doesn't attempt to convert it to an UNMAP. 

Since RBD and Gluster at the only two plugins that implement write-same, maybe it would be cleaner to just expose a direct hook w/ the `unmap` boolean and then let those two plugins attach to the hook directly instead of intercepting it in `XYZ_handle_cmd` and expecting an emulation-like callback from `tcmur_handle_writesame`

Reply:
> `not aligns` -> `not aligned` (also probably not a warning since it's only an issue because `tcmu` is not treating the WS as a WS). 

Sure, will fix it.

> Looking at the kernel's `drivers/scsi/sd.c`, it translates a zero extent call to a WS+UNMAP and doesn't attempt to convert it to an UNMAP.
> 
Checked it, it just translate the zero request to WA+UNMAP SCSI command, which the hardware could handle. 

The scsi disk hardware controller, which will be equivalent to tcmu-runner/librbd, will decide how to do if it receives a WS command with UNMAP bit set and not aligned. I guess, the scsi disk will just write zero to the segment not aligned. If all are aligned, the UNMAP should be the most efficient way to zero the disk. So internally it will switch to UNMAP I think. This is why the tcmu-runner is trying to switch to UNMAP if WS have explicitly allow it.

This is what the `drivers/target/target_core_sbc.c` will do.

Reply:
> Since RBD and Gluster at the only two plugins that implement write-same, maybe it would be cleaner to just expose a direct hook w/ the `unmap` boolean and then let those two plugins attach to the hook directly instead of intercepting it in `XYZ_handle_cmd` and expecting an emulation-like callback from `tcmur_handle_writesame`

Yeah, looks better. Let's expose the handler->writesame and handler->caw hooks to clean up the code.